packages/nimble-components/src/table/index.ts

```diff
@@ -9,8 +9,17 @@ import {
     getCoreRowModel as tanStackGetCoreRowModel,
     TableOptionsResolved as TanStackTableOptionsResolved
 } from '@tanstack/table-core';
+import {
+    Virtualizer,
+    VirtualizerOptions,
+    elementScroll,
+    observeElementOffset,
+    observeElementRect,
+    VirtualItem
+} from '@tanstack/virtual-core';
```

Comment:
I can't recall if we talked about this during the spec process, but one of the side effects of virtualization is that transient state isn't automatically maintained if you scroll away. One symptom of this that I can see right now is if you select the text of a row and then scroll, the selection jumps around to other rows.

We may need to brainstorm other state for future column types. Text cursor position in editable cells is the only one that comes to mind right now.

Thoughts on how hard it would be to include this state when we configure cells to be displayed?

https://user-images.githubusercontent.com/10500124/213286827-ceec2749-d16d-4f99-a30b-ae97280c68a5.mov

Reply:
This'll probably need some discussion (I'd like to hear @atmgrifter00 's thoughts).

My thinking was that when a scroll is detected (virtualizer onChange), before we change the visible items (which re-renders the rows as you scroll), we'd give the old rows a chance to commit edits / close popups / etc, and we'd need need some code to deselect any text that was selected. But I'm not sure that we need to restore any state when you returned to the previous rows/cells (I think that's what you're suggesting anyway).

Reply:
Yeah, I am suggesting that users would likely expect that state to be maintained while the cell remains in view and also if it's scrolled offscreen and then back into view.

But I could easily see that being a challenge to implement and possibly not a showstopper if we don't restore the state so wanted to start a discussion about the cost of various options.

Sounds like we have 3 options:
1. leave as-is. I don't think this desirable because state appears to jump to different cells as you scroll. That said, it looks like our existing grid has the same bug with selection.
2. clear state on scroll. Better, but probably users would be surprised that selection goes away as soon as you start scrolling. It prevents the power user workflow of selecting some rows to "bookmark" them to find them quickly after scrolling away
3. persist state on scroll but restore it before cells are rendered. This would be best because it preserves the illusion that all the cells always exist. But we'd have to cover corner cases like text selection changing while a selected row is offscreen or a selected row being deleted from the data.

We may also pick different answers for different types of state (selection vs active edits vs open popups).

I'm curious for others opinions too, on both the cost and desirability of various options.

Reply:
I think we can also potentially delineate preserving row state on scroll, and cell state. I, for one, don't see much of an issue with not preserving cell state on scroll. For the most part, at least in the known immediate known usages of the table, the only real cell state of concern is open popups, and I think it is totally acceptable to dismiss popups on scroll.

I think, ultimately, we definitely want to preserve row state while scrolling. Currently, there is no row state of consequence that we need to preserve, but once we have row selected we obviously will.

So, at the moment, I am in favor of doing the set of tasks Malcolm outlines for "unloading" any current cell state once a scroll is initiated. For the row state preservation, I would think we could tackle this once we introduce behaviors that involve persisted row state like selection.

Reply:
Recording some of the state we mentioned while brainstorming this in standup today:
1. text selection. This can span multiple rows
2. input focus (where is the text cursor)
3. pending text input (text you've typed but not committed)
4. open popups (action menu or select/combobox cells whose popup is open)

For each of these we need to decide whether to manually clear it on scroll and whether to restore it on scroll.

I agree that row selection is in a different category which must be preserved as you scroll (but hopefully that's easy thanks to tanstack tracking it)

Reply:
It seems related but could be treated as a separate issue, but on android I see that when you touch a row it activates the hover state ([chromium issue](https://bugs.chromium.org/p/chromium/issues/detail?id=370155)), and if you then scroll you see the "hovered" row move around.

I was able to reproduce it in devtools with device mode turned on:

![hovereffect2](https://user-images.githubusercontent.com/1588923/214101574-e3cf43eb-0529-4c3a-b807-b06ea6f02c89.gif)

Reply:
As mentioned in one of our standups, my plan is to spin these issues off into a follow-up work item.

(Leaving this comment thread open until I create that work item)

Reply:
Moved/Split to #997